src/http1.cr

```
@@ -0,0 +1,409 @@
+        while slice.first? == SP
+          slice += 1
+        end
+
+        # HTTP-version
+        space_index = slice.index(SP) || slice.size
+        case slice[0, space_index]
+        when "HTTP/1.0".to_slice
+          @version = "HTTP/1.0"
+        when "HTTP/1.1".to_slice
+          @version = "HTTP/1.1"
+        else
+          return HTTP::Status::BAD_REQUEST
+        end
+        slice += space_index
+
+        # SP* EOL
+        while byte = slice.first?
+          return HTTP::Status::BAD_REQUEST unless byte == SP
+          slice += 1
+        end
+
+        {method, path}
+      end
+    end
+
+    def read_status_line : {String, Int32, String}
+      read_line(max_status_line_size, nil) do |slice|
+        # status-line = HTTP-version SP+ status SP+ reason-phrase SP* [CR] LF
+        space_index = slice.index(SP)
+        raise "Invalid status line" unless space_index
+
+        version =
+          case subslice = slice[0, space_index]
+          when "HTTP/1.0".to_slice
+            "HTTP/1.0"
+          when "HTTP/1.1".to_slice
+            "HTTP/1.1"
+          else
+            raise "Unsupported HTTP version #{String.new(subslice)}"
+          end
+        slice += space_index
+
+        # Per RFC 9112, Section 4. there MUST be a single SP but implementations
+        # MAY skip any whitespace character (SP, HTAB, VT, FF and a lone CR).
+        # Here we merely skip multiple SP.
+        while slice.first? == SP
+          slice += 1
+        end
+
+        # status (3 digits)
+        space_index = slice.index(SP) || slice.size
+        subslice = slice[0, space_index]
+        status = subslice.reduce(0) do |acc, byte|
+          case byte
+          when ('0'.ord)..('9'.ord)
+            acc * 10 + (byte - '0'.ord)
+          else
+            raise "Invalid HTTP status code: #{String.new(subslice)}"
+          end
+        end
+        raise "Invalid HTTP status code: #{status}" unless 100 <= status <= 999
+        slice += 3
```

The loop could be reworked:
- We can fail early when `space_index` isn't 3 to avoid an inexplicit math overflow;
- We can consume the bytes out of `slice` directly (no need for subslice).